File: p316/p316.cpp
// O(N). Nearly brutal..
class Solution {
public:
    string removeDuplicateLetters(string s) {
        int n = s.length(), tot = 0;

        vector<bool> vis(26);
        vector<int> count(26);

        for (int i = n-1; i >= 0; --i) {
        	int c = s[i] - 'a';
        	count[c] = tot;

        	if (!vis[c]) {
        		vis[c] = true;
        		++tot;
        	}
        }

        string res(tot, 'a');

        for (int i = 0; i < tot; ++i) {
        	char best;
        	int maxCount = 0;

        	for (int j = 0; j < 26; ++j)
        		if (vis[j] && count[j] > maxCount) {
        			best = 'a' + j;
        			maxCount = count[j];
        		}

        	res[i] = best;
        	vis[best-'a'] = false;

        	if (i == tot-1)
        		break;


        	count = vector<int>(26, 0);
        	vector<bool> seen(26);

        	int sum = 0;
        	for (int j = n-1; j >= 0; --j) {
        		int c = s[i] - 'a';
        		if (vis[c]) {
        			count[i] = sum;

        			if (!seen[c]) {
        				seen[c] = true;
        				++sum;
        			}
        		}
        	}
        }

        return res;
    }
};

File: p316/p316.go
package p316

func removeDuplicateLetters(s string) string {
	n := len(s)

	st := []byte{}
	count := make(map[byte]int)
	used := make(map[byte]bool)

	for i := 0; i < n; i++ {
		count[s[i]]++
	}

	for i := 0; i < n; i++ {
		count[s[i]]--

		if used[s[i]] {
			continue
		}

		for len(st) > 0 && st[len(st)-1] > s[i] && count[st[len(st)-1]] > 0 {
			used[st[len(st)-1]] = false
			st = st[:len(st)-1]
		}

		used[s[i]] = true
		st = append(st, s[i])
	}

	return string(st)
}
